salt/utils/files.py

```diff
@@ -547,6 +547,8 @@ def _onerror(func, path, exc_info):
     if os.path.islink(path) or not os.path.isdir(path):
         os.remove(path)
     else:
+        if six.PY2 and salt.utils.platform.is_windows() and isinstance(path, six.string_types):
+            path = path.decode('utf-8')
```

Comment:
`salt.utils.stringutils.to_unicode()` should be used here instead. The reason for this is that the `unicode` type's `decode()` method will attempt to encode the string before decoding, because reasons:

```
>>> u'Д'.decode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.7/encodings/utf_8.py", line 16, in decode
    return codecs.utf_8_decode(input, errors, True)
UnicodeEncodeError: 'ascii' codec can't encode character u'\u0414' in position 0: ordinal not in range(128)
```

We can skip the PY2 check completely because the `to_unicode()` helper already performs this check. So, these two lines can be alternately written as:

```python
if salt.utils.platform.is_windows():
    try:
        path = salt.utils.stringutils.to_unicode(path)
    except TypeError:
        pass
```

If `to_unicode()` is passed a non string/bytestring/bytearray type, it will raise a `TypeError`, hence the exception catching. This also allows us to avoid the `isinstance` check.